File: frontend/tests/lib/application-status.test.ts
import {
  adminNextStatuses,
  canUserTransition,
  isTerminalStatus,
  nextActionLabel,
  normalizeApplicationStatus,
  userNextStatuses,
} from '@/lib/application-status'

describe('application-status', () => {
  it('applied から accepted への直接遷移は拒否する', () => {
    expect(canUserTransition('applied', 'accepted')).toBe(false)
    expect(userNextStatuses('applied')).toEqual(['withdrawn'])
  })

  it('offered からは承諾と辞退だけ選べる', () => {
    expect(userNextStatuses('offered')).toEqual(['accepted', 'withdrawn'])
    expect(canUserTransition('offered', 'accepted')).toBe(true)
  })

  it('終了状態からの再遷移を制限する', () => {
    expect(isTerminalStatus('accepted')).toBe(true)
    expect(isTerminalStatus('declined')).toBe(true)
    expect(userNextStatuses('accepted')).toEqual([])
    expect(canUserTransition('rejected', 'applied')).toBe(false)
  })

  it('レガシー interview / declined を現行コードへ正規化する', () => {
    expect(normalizeApplicationStatus('interview')).toBe('interview_in_progress')
    expect(normalizeApplicationStatus('declined')).toBe('withdrawn')
    expect(userNextStatuses('interview')).toEqual(['withdrawn'])
  })
})

describe('adminNextStatuses', () => {
  it('applied から書類選考と辞退へ進める', () => {
    expect(adminNextStatuses('applied')).toEqual(['document_screening', 'withdrawn'])
  })

  it('レガシー interview は面接中として次ステータスを返す', () => {
    expect(adminNextStatuses('interview')).toEqual(['offered', 'rejected', 'withdrawn'])
  })
})

describe('nextActionLabel', () => {
  it('選べる操作が1つなら「◯◯する」の具体的なラベルにする', () => {
    expect(nextActionLabel('applied')).toBe('辞退する')
    expect(nextActionLabel('interview_scheduled')).toBe('辞退する')
  })

  it('選べる操作が複数なら選択肢を列挙する', () => {
    expect(nextActionLabel('offered')).toBe('内定承諾・辞退を選ぶ')
  })

  it('終了状態など次の操作がない場合は空文字を返す', () => {
    expect(nextActionLabel('accepted')).toBe('')
    expect(nextActionLabel('rejected')).toBe('')
  })
})

File: Backend/internal/models/application_status_test.go
package models

import (
	"slices"
	"strings"
	"testing"
)

func TestFlywheelPassedStatuses_IncludesDocumentPassedOnward(t *testing.T) {
	want := []string{
		"document_passed",
		"interview_scheduled",
		"interview_in_progress",
		"offered",
		"accepted",
	}
	for _, s := range want {
		if !slices.Contains(FlywheelPassedStatuses, s) {
			t.Errorf("FlywheelPassedStatuses に %s が含まれていない", s)
		}
	}
}

func TestFlywheelPassedStatuses_ExcludesNonPass(t *testing.T) {
	notPassed := []string{
		"not_applied", "applied", "document_screening",
		"withdrawn", "rejected", "declined", "interview",
	}
	for _, s := range notPassed {
		if slices.Contains(FlywheelPassedStatuses, s) {
			t.Errorf("FlywheelPassedStatuses に非通過 %s が含まれている", s)
		}
	}
}

func TestFlywheelPassedStatusFilter_IncludesLegacyInterview(t *testing.T) {
	filter := FlywheelPassedStatusFilter()
	if !slices.Contains(filter, "interview") {
		t.Error("旧 interview が読み取りフィルタに含まれていない")
	}
	if !slices.Contains(filter, "interview_in_progress") {
		t.Error("interview_in_progress が読み取りフィルタに含まれていない")
	}
	for _, s := range []string{"withdrawn", "rejected", "declined"} {
		if slices.Contains(filter, s) {
			t.Errorf("フィルタに非通過 %s が含まれている", s)
		}
	}
}

func TestFlywheelPassedStatusSQLIn_MatchesFilter(t *testing.T) {
	got := FlywheelPassedStatusSQLIn()
	for _, s := range FlywheelPassedStatusFilter() {
		if !strings.Contains(got, "'"+s+"'") {
			t.Errorf("SQL IN に '%s' がない: %s", s, got)
		}
	}
	for _, s := range []string{"withdrawn", "rejected", "declined"} {
		if strings.Contains(got, "'"+s+"'") {
			t.Errorf("SQL IN に非通過 '%s' がある: %s", s, got)
		}
	}
}
